core/tee/tee_svc.c

```diff
@@ -616,6 +616,13 @@ static TEE_Result tee_svc_copy_param(struct tee_ta_session *sess,
 			if (!va) {
 				if (s)
 					return TEE_ERROR_BAD_PARAMETERS;
+
+				/*
+				 * Ensure mobj is NULL because
+				 * utee_param_to_param set it to mobj_virt
+				 * even if there is not user buffer
+				 */
+				param->u[n].mem.mobj = NULL;
```

Comment:
How about fixing this in `utee_param_to_param()` instead? The root of the problem seems to be there.

Reply:
Do you accept the modification of the utee_param_to_param by adding a if (!a && !b) then mobj = NULL? (in case of IN/OUT Buffer for sure)

Reply:
I was thinking of something like:
``` C
p->u[n].mem.offs = a;
p->u[n].mem.size = b;
if (!p->u[n].mem.offs) {
    /* Allow NULL memrefs, but they have to be of size 0 */
    if (p->u[n].mem.size)
         return TEE_ERROR_BAD_PARAMETERS;
     p->u[n].mem.mobj = NULL;
     break;
}
p->u[n].mem.mobj = &mobj_virt;
if (tee_mmu_check_access_rights...
```

Reply:
I'm concerned with what's done in **libteec**.

For allocated shm:
https://github.com/OP-TEE/optee_client/blob/3.8.0/libteec/src/tee_client_api.c#L745-L752
Null size allocation ends in `malloc(0)` which may or may not return a NULL buffer, depending on the OS. We don't expect to rely on the OS behavior.

For registered shm:
https://github.com/OP-TEE/optee_client/blob/3.8.0/libteec/src/tee_client_api.c#L677-L679
If the registered shm was allocated from `malloc(0)`, then behavior also depends on the OS, hence out of OP-TEE client/OS control.

Won't this jeopardize the NULL_MEMREF feature?

Reply:
I've edited my previous comment.
For allocated SHM, seems no issue since we do have a non NULL buffer.
For registered SHM, I wonder if registering a NULL buffer with a non zero size is a good thing.

Reply:
@etienne-lms, I agree. It seems that the `TEEC_RegisterSharedMemory()` isn't correct.
Perhaps we can start a new conversation in a PR or Issue in https://github.com/OP-TEE/optee_client/.